models/schemas/levelSchema.ts

```diff
@@ -1,60 +1,112 @@
+import { LevelModel, UserModel } from '../mongoose';
 import Level from '../db/level';
 import mongoose from 'mongoose';
 
-const LevelSchema = new mongoose.Schema<Level>({
-  _id: {
-    type: mongoose.Schema.Types.ObjectId,
-    required: true,
-  },
-  authorNote: {
-    type: String,
-  },
-  // data format is a string of 'LevelDataType's with rows separated by '\n'
-  data: {
-    type: String,
-    required: true,
-  },
-  height: {
-    type: Number,
-    required: true,
-  },
-  isDraft: {
-    type: Boolean,
-    required: true,
-  },
-  leastMoves: {
-    type: Number,
-    required: true,
-  },
-  name: {
-    type: String,
-    required: true,
-  },
-  points: {
-    type: Number,
-    required: true,
-  },
-  psychopathId: {
-    type: Number,
-  },
-  ts: {
-    type: Number,
-    required: true,
-  },
-  userId: {
-    type: mongoose.Schema.Types.ObjectId,
-    ref: 'User',
-    required: true,
-  },
-  width: {
-    type: Number,
-    required: true,
-  },
-}, {
-  collation: {
-    locale: 'en_US',
-    strength: 2,
+const LevelSchema = new mongoose.Schema<Level>(
+  {
+    _id: {
+      type: mongoose.Schema.Types.ObjectId,
+      required: true,
+    },
+    authorNote: {
+      type: String,
+    },
+    // data format is a string of 'LevelDataType's with rows separated by '\n'
+    data: {
+      type: String,
+      required: true,
+    },
+    height: {
+      type: Number,
+      required: true,
+    },
+    isDraft: {
+      type: Boolean,
+      required: true,
+    },
+    leastMoves: {
+      type: Number,
+      required: true,
+    },
+    name: {
+      type: String,
+      required: true,
+    },
+    points: {
+      type: Number,
+      required: true,
+    },
+    psychopathId: {
+      type: Number,
+    },
+    slug: {
+      type: String,
+      slug: 'name',
+    },
+    ts: {
+      type: Number,
+      required: true,
+    },
+    userId: {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: 'User',
+      required: true,
+    },
+    width: {
+      type: Number,
+      required: true,
+    },
   },
+  {
+    collation: {
+      locale: 'en_US',
+      strength: 2,
+    },
+  }
+);
+
+LevelSchema.index({ slug: 1 }, { name: 'slug_index' });
+
+LevelSchema.pre('save', function (next) {
```

Comment:
Do we need to generate all slugs in advance before pushing this PR, since slugs are only added on update?

Reply:
Or we can generate them on the fly when a level is requested and slug for the level is not found.. But probably easier to just update them all